solr/core/src/java/org/apache/solr/update/processor/DistributedZkUpdateProcessor.java

```diff
@@ -184,6 +186,30 @@ public void processCommit(CommitUpdateCommand cmd) throws IOException {
 
     updateCommand = cmd;
 
+    // 1. SHARED replica has a hard requirement of processing each indexing batch with a hard commit(either explicit or
+    // implicit, HttpSolrCall#addCommitIfAbsent) because that is how, at the end of an indexing batch, synchronous push
+    // to shared store gets hold of the segment files on local disk. SHARED replica also does not support the notion of soft commit.
+    // Therefore unlike NRT replica type we do not need to broadcast commit to the leaders of all the shards of a collection.
+    //
+    // 2. <code>isLeader</code> is computed fresh each time an AddUpdateCommand/DeleteUpdateCommand belonging to the indexing
+    // batch is processed. And finally it is recomputed in this method. It is possible that at the beginning of a batch
+    // this replica was a leader and did process some AddUpdateCommand/DeleteUpdateCommand. But before reaching this 
+    // method lost the leadership. In that case we will still like to process the commit otherwise the indexing batch can
+    // succeed without pushing the changes to the shared store (data loss). Therefore, we are not restricting the 
```

Comment:
SHARED replica does not need leadership as such because it relies on the optimistic concurrency when writing to metadataSuffix znode. As long as a replica can match the metadataSuffix version it started indexing with, it will be correct. If the new leader has started indexing and have pushed before this replica then this replica will fail. But if it pushes before the new leader can push then the batch on new leader will fail. In both cases indexing will be correct.